lgc/patch/RegisterMetadataBuilder.cpp

```diff
@@ -0,0 +1,1380 @@
+
+  assert(calcFactor.primAmpFactor >= 1);
+  unsigned maxVertsPerSubgroup = NggMaxThreadsPerSubgroup;
+  unsigned threadsPerSubgroup = NggMaxThreadsPerSubgroup;
+  unsigned spiShaderIdsFormat = SPI_SHADER_1COMP;
+  if (m_hasMesh) {
+    maxVertsPerSubgroup = std::min(meshMode.outputVertices, NggMaxThreadsPerSubgroup);
+    threadsPerSubgroup = calcFactor.primAmpFactor;
+    const bool enableMultiView = m_pipelineState->getInputAssemblyState().enableMultiView;
+    bool hasPrimitivePayload = meshBuiltInUsage.layer || meshBuiltInUsage.viewportIndex ||
+                               meshBuiltInUsage.primitiveShadingRate || enableMultiView;
+    if (m_gfxIp.major < 11)
+      hasPrimitivePayload |= meshBuiltInUsage.primitiveId;
+    spiShaderIdsFormat = hasPrimitivePayload ? SPI_SHADER_2COMP : SPI_SHADER_1COMP;
+
+    // VGT_DRAW_PAYLOAD_CNTL
+    getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VgtDrawPrimPayloadEn] = hasPrimitivePayload;
+
+    if (m_gfxIp.major >= 11) {
+      // SPI_SHADER_GS_MESHLET_DIM
+      auto spiShaderGsMeshletDim =
+          getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::SpiShaderGsMeshletDim].getMap(true);
+      spiShaderGsMeshletDim[Util::Abi::SpiShaderGsMeshletDimMetadataKey::NumThreadX] = meshMode.workgroupSizeX - 1;
+      spiShaderGsMeshletDim[Util::Abi::SpiShaderGsMeshletDimMetadataKey::NumThreadY] = meshMode.workgroupSizeY - 1;
+      spiShaderGsMeshletDim[Util::Abi::SpiShaderGsMeshletDimMetadataKey::NumThreadZ] = meshMode.workgroupSizeZ - 1;
+      // NOTE: If row export for mesh shader is enabled, the thread group size is set according to dimensions of work
+      // group. Otherwise, it is set according to actual primitive amplification factor.
+      const unsigned threadGroupSize = m_pipelineState->enableMeshRowExport()
+                                           ? meshMode.workgroupSizeX * meshMode.workgroupSizeY * meshMode.workgroupSizeZ
+                                           : calcFactor.primAmpFactor;
+      spiShaderGsMeshletDim[Util::Abi::SpiShaderGsMeshletDimMetadataKey::ThreadgroupSize] = threadGroupSize;
+
+      // SPI_SHADER_GS_MESHLET_EXP_ALLOC
+      auto spiShaderGsMeshletExpAlloc =
+          getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::SpiShaderGsMeshletExpAlloc].getMap(true);
+      spiShaderGsMeshletExpAlloc[Util::Abi::SpiShaderGsMeshletExpAllocMetadataKey::MaxExpVerts] =
+          meshMode.outputVertices;
+      spiShaderGsMeshletExpAlloc[Util::Abi::SpiShaderGsMeshletExpAllocMetadataKey::MaxExpPrims] =
+          meshMode.outputPrimitives;
+    }
+  } else {
+    maxVertsPerSubgroup = std::min(gsInstPrimsInSubgrp * maxVertOut, NggMaxThreadsPerSubgroup);
+    // VGT_GS_VERT_ITEMSIZE
+    unsigned gsVertItemSizes[4] = {4 * gsInOutUsage.outputMapLocCount, 0, 0, 0};
+    getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VgtGsVertItemsize] = gsVertItemSizes;
```

Comment:
Another "address of array will always evaluate to true".

Reply:
Fixed in [d061491](https://github.com/GPUOpen-Drivers/llpc/commit/d0614913041bd72c1f3999d373fb19323ee72b2d).